### PR TITLE
Add public API

### PR DESCRIPTION
**Pull request by fatlotus**
Rudimentary documentation is here:
https://docs.google.com/document/d/1uHTq_U0_v97KuHV1LElDp3hNwrcykaPrk61kT2Igti4/edit#

Thoughts? After our meeting today @georgeteo, I think we agreed that this was not a huge priority... but then again we've had a few people asking.

### CODE REVIEW

File: caravel/controllers/api.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    """Returns a safe JSON blob representing the given listing."""
+
+    return {
+        "jsonURL": url_for(
+            "api_one_listing", permalink=listing.permalink, _external=True),
+        "htmlURL": url_for(
+            "show_listing", permalink=listing.permalink, _external=True),
+        "title": listing.title,
+        "body": listing.body,
+        "postingTimeInSSince1970": listing.posting_time,
```
Comment (georgeteo) [inline caravel/controllers/api.py]: postingTimeInSSince1970?

File: caravel/controllers/api.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import uuid, time
+
+from flask import render_template, request, abort, redirect, url_for, session
+from flask import flash, g, jsonify
+
+from caravel import app
+from caravel.storage import helpers, entities, photos, dos
+
+def _externalize(listing):
```
Comment (georgeteo) [inline caravel/controllers/api.py]: Is better to store/cache this somewhere the recomputing for each api call?

File: caravel/controllers/api.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                    "1uHTq_U0_v97KuHV1LElDp3hNwrcykaPrk61kT2Igti4/edit")
+
+@app.route("/api/v1/listings.json")
+def api_all_listings():
+    """Display a list of listings that match the given query."""
+
+    # Parse filtering options from query.
+    query = request.args.get("q", "")
+    offset = int(request.args.get("offset", "0"))
+    if offset < 0:
```
Comment (georgeteo) [inline caravel/controllers/api.py]: Will offset be less than 0?

File: requirements.txt [hunk 4]
```diff
@@ -1,7 +0,0 @@
-flask_bootstrap
```
Comment (georgeteo) [inline requirements.txt]: Why did you remove this?

File: caravel/controllers/api.py (hunk 2 above)
Comment (fatlotus) [inline caravel/controllers/api.py]: Shouldn't be too slow — what parts are you worried about?

File: caravel/controllers/api.py (hunk 1 above)
Comment (fatlotus) [inline caravel/controllers/api.py]: fixed

File: caravel/controllers/api.py (hunk 3 above)
Comment (fatlotus) [inline caravel/controllers/api.py]: Shouldn't be — but this is a way to validate that it isn't :p
